util/regexp-assemble/regexp-assemble.py

```diff
@@ -9,140 +9,216 @@
+        help="The six digit ID of the rule, e.g. 932100",
+    )
+    parser.add_argument(
+        "--all",
+        action="store_true",
+        help="""
+    Instead of supplying a rule_id, you can tell the script to
+    compare all regular expressions
+    """,
+    )
+    parser.set_defaults(func=handle_compare)
+
+
+def handle_generate(namespace: argparse.Namespace):
+    context = create_context()
+    assembler = Assembler(context)
+
+    if namespace.rule_id:
+        with open(
+            os.path.join(context.data_files_directory, f"{namespace.rule_id}.data"),
+            "rt",
+        ) as handle:
+            regex = assembler.run(handle)
+    elif "stdin" in namespace:
+        regex = assembler.run(namespace.stdin)
+    else:
+        raise argparse.ArgumentError("Unhandled combination of arguments")
+
+    sys.stdout.write(regex)
+    sys.stdout.write("\n")
+
+
+def handle_update(namespace: argparse.Namespace):
+    updater = Updater(create_context())
+    if namespace.rule_id:
+        updater.run(namespace.rule_id)
+    elif namespace.all:
+        updater.run_all()
+
+
+def handle_compare(namespace: argparse.Namespace):
+    comparer = Comparer(create_context())
+    if namespace.rule_id:
+        comparer.run(namespace.rule_id)
+    elif namespace.all:
+        comparer.run_all()
+
+
+class RuleNameParser(argparse.Action):
+    def __init__(
+        self,
+        option_strings,
+        dest,
+        nargs=None,
+        const=None,
+        default=None,
+        type=None,
+        choices=None,
+        required=False,
+        help=None,
+        metavar=None,
+    ):
+        super().__init__(
+            option_strings,
+            dest,
+            nargs=nargs,
+            const=const,
+            default=default,
+            type=type,
+            choices=choices,
+            required=required,
+            help=help,
+            metavar=metavar,
+        )
+
+    def __call__(
+        self,
+        parser: argparse.ArgumentParser,
+        namespace: argparse.Namespace,
+        values: list,
+        option_string: str,
+    ):
+        if values == "-":
+            setattr(namespace, "stdin", sys.stdin)
+            return
+        elif not values:
+            if self.nargs == "?" and not namespace.all:
+                raise ValueError("Invalid arguments")
+            return
+
+        match = re.fullmatch(r"(\d{6})(?:\.data)?", values)
+        if not match:
+            raise ValueError(f"Failed to identify rule from argument {values}")
+        setattr(namespace, self.dest, match.group(1))
+
+
+def create_context() -> Context:
+    script_directory = os.path.dirname(__file__)
+    root_directory = os.path.dirname(os.path.dirname(script_directory))
+    return Context(root_directory)
+
+
+def setup_logger(namespace: argparse.Namespace):
+    logger = logging.getLogger()
+    logger.setLevel(namespace.log_level)
+    logger.addHandler(logging.StreamHandler())
+
+
+if __name__ == "__main__":
```

Comment:
I think some error handling would be fine here, eg. if the sub command is missing, or the mandatory argument. Eg.:

* run the tool without sub command:
```
$ util/regexp-assemble/regexp-assemble.py
Traceback (most recent call last):
  File "/home/airween/virtualenv/coreruleset/util/regexp-assemble/regexp-assemble.py", line 224, in <module>
    namespace.func(namespace)
AttributeError: 'Namespace' object has no attribute 'func'
```
* run the tool with sub command but without mandatory argument:
```
$ util/regexp-assemble/regexp-assemble.py compare
Traceback (most recent call last):
  File "/home/airween/virtualenv/coreruleset/util/regexp-assemble/regexp-assemble.py", line 222, in <module>
    namespace = build_args_parser().parse_args()
  File "/usr/lib/python3.9/argparse.py", line 1830, in parse_args
    args, argv = self.parse_known_args(args, namespace)
  File "/usr/lib/python3.9/argparse.py", line 1863, in parse_known_args
    namespace, args = self._parse_known_args(args, namespace)
  File "/usr/lib/python3.9/argparse.py", line 2075, in _parse_known_args
    stop_index = consume_positionals(start_index)
  File "/usr/lib/python3.9/argparse.py", line 2031, in consume_positionals
    take_action(action, args)
  File "/usr/lib/python3.9/argparse.py", line 1940, in take_action
    action(self, namespace, argument_values, option_string)
  File "/usr/lib/python3.9/argparse.py", line 1219, in __call__
    subnamespace, arg_strings = parser.parse_known_args(arg_strings, None)
  File "/usr/lib/python3.9/argparse.py", line 1863, in parse_known_args
    namespace, args = self._parse_known_args(args, namespace)
  File "/usr/lib/python3.9/argparse.py", line 2075, in _parse_known_args
    stop_index = consume_positionals(start_index)
  File "/usr/lib/python3.9/argparse.py", line 2031, in consume_positionals
    take_action(action, args)
  File "/usr/lib/python3.9/argparse.py", line 1940, in take_action
    action(self, namespace, argument_values, option_string)
  File "/home/airween/virtualenv/coreruleset/util/regexp-assemble/regexp-assemble.py", line 200, in __call__
    raise ValueError("Invalid arguments")
ValueError: Invalid arguments

```

Reply:
Done.